File: ambari-metrics/ambari-metrics-alertservice/src/main/java/org/apache/ambari/metrics/alertservice/spark/MetricAnomalyDetector.java
package org.apache.ambari.metrics.alertservice.spark;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.ambari.metrics.alertservice.common.MetricAnomaly;
import org.apache.ambari.metrics.alertservice.common.TimelineMetric;
import org.apache.ambari.metrics.alertservice.common.TimelineMetrics;
import org.apache.ambari.metrics.alertservice.methods.ema.EmaModel;
import org.apache.ambari.metrics.alertservice.methods.MetricAnomalyModel;
import org.apache.ambari.metrics.alertservice.methods.ema.EmaModelLoader;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.streaming.Duration;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaPairReceiverInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.kafka.KafkaUtils;
import scala.Tuple2;

import java.util.*;

public class MetricAnomalyDetector {

    private static final Log LOG = LogFactory.getLog(MetricAnomalyDetector.class);
    private static String groupId = "ambari-metrics-group";
    private static String topicName = "ambari-metrics-topic";
    private static int numThreads = 1;

    //private static String zkQuorum = "avijayan-ams-1.openstacklocal:2181,avijayan-ams-2.openstacklocal:2181,avijayan-ams-3.openstacklocal:2181";
    //private static Map<String, String> kafkaParams = new HashMap<>();
    //static {
    //    kafkaParams.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "avijayan-ams-2.openstacklocal:6667");
    //    kafkaParams.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer");
    //    kafkaParams.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.connect.json.JsonSerializer");
    //    kafkaParams.put("metadata.broker.list", "avijayan-ams-2.openstacklocal:6667");
    //}

    public MetricAnomalyDetector() {
    }

    public static void main(String[] args) throws InterruptedException {


        if (args.length < 6) {
            System.err.println("Usage: MetricAnomalyDetector <method1,method2> <appid1,appid2> <collector_host> <port> <protocol> <zkQuorum>");
            System.exit(1);
        }

        List<String> appIds = Arrays.asList(args[1].split(","));
        String collectorHost = args[2];
        String collectorPort = args[3];
        String collectorProtocol = args[4];
        String zkQuorum = args[5];

        List<MetricAnomalyModel> anomalyDetectionModels = new ArrayList<>();
        AnomalyMetricPublisher anomalyMetricPublisher = new AnomalyMetricPublisher(collectorHost, collectorProtocol, collectorPort);

        SparkConf sparkConf = new SparkConf().setAppName("AmbariMetricsAnomalyDetector");

        JavaStreamingContext jssc = new JavaStreamingContext(sparkConf, new Duration(10000));

        for (String method : args[0].split(",")) {
            if (method.equals("ema")) {
                LOG.info("Model EMA requested.");
                EmaModel emaModel = new EmaModelLoader().load(jssc.sparkContext().sc(), "/tmp/model/ema");
                anomalyDetectionModels.add(emaModel);
            }
        }

        JavaPairReceiverInputDStream<String, String> messages =
                KafkaUtils.createStream(jssc, zkQuorum, groupId, Collections.singletonMap(topicName, numThreads));

        //Convert JSON string to TimelineMetrics.
        JavaDStream<TimelineMetrics> timelineMetricsStream = messages.map(new Function<Tuple2<String, String>, TimelineMetrics>() {
            @Override
            public TimelineMetrics call(Tuple2<String, String> message) throws Exception {
                LOG.info(message._2());
                ObjectMapper mapper = new ObjectMapper();
                TimelineMetrics metrics = mapper.readValue(message._2, TimelineMetrics.class);
                return metrics;
            }
        });

        //Group TimelineMetric by AppId.
        JavaPairDStream<String, TimelineMetrics> appMetricStream = timelineMetricsStream.mapToPair(
                timelineMetrics -> new Tuple2<String, TimelineMetrics>(timelineMetrics.getMetrics().get(0).getAppId(),timelineMetrics)
        );

        appMetricStream.print();

        //Filter AppIds that are not needed.
        JavaPairDStream<String, TimelineMetrics> filteredAppMetricStream = appMetricStream.filter(new Function<Tuple2<String, TimelineMetrics>, Boolean>() {
            @Override
            public Boolean call(Tuple2<String, TimelineMetrics> appMetricTuple) throws Exception {
                return appIds.contains(appMetricTuple._1);
            }
        });

        filteredAppMetricStream.print();

        filteredAppMetricStream.foreachRDD(rdd -> {
            rdd.foreach(
                    tuple2 -> {
                        TimelineMetrics metrics = tuple2._2();
                        LOG.info("Received Metric : " + metrics.getMetrics().get(0).getMetricName());
                        for (TimelineMetric metric : metrics.getMetrics()) {

                            TimelineMetric timelineMetric =
                                    new TimelineMetric(metric.getMetricName(), metric.getAppId(), metric.getHostName(), metric.getMetricValues());
                            LOG.info("Models size : " + anomalyDetectionModels.size());

                            for (MetricAnomalyModel model : anomalyDetectionModels) {
                                LOG.info("Testing against Model : " + model.getClass().getCanonicalName());
                                List<MetricAnomaly> anomalies = model.test(timelineMetric);
                                anomalyMetricPublisher.publish(anomalies);
                                for (MetricAnomaly anomaly : anomalies) {
                                    LOG.info(anomaly.getAnomalyAsString());
                                }

                            }
                        }
                    });
        });

        jssc.start();
        jssc.awaitTermination();
    }
}

File: ambari-metrics/ambari-metrics-spark/src/main/scala/org/apache/ambari/metrics/spark/MetricAnomalyDetector.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ambari.metrics.spark


import java.util
import java.util.logging.LogManager

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.spark.SparkConf
import org.apache.spark.streaming._
import org.apache.spark.streaming.kafka._
import org.apache.ambari.metrics.alertservice.common.{MetricAnomaly, TimelineMetrics}
import org.apache.ambari.metrics.alertservice.methods.MetricAnomalyModel
import org.apache.ambari.metrics.alertservice.methods.ema.{EmaModel, EmaModelLoader}
import org.apache.ambari.metrics.alertservice.spark.AnomalyMetricPublisher
import org.apache.log4j.Logger
import org.apache.spark.storage.StorageLevel

import scala.collection.JavaConversions._
import org.apache.logging.log4j.scala.Logging

object MetricAnomalyDetector extends Logging {


  var zkQuorum = "avijayan-ams-1.openstacklocal:2181,avijayan-ams-2.openstacklocal:2181,avijayan-ams-3.openstacklocal:2181"
  var groupId = "ambari-metrics-group"
  var topicName = "ambari-metrics-topic"
  var numThreads = 1
  val anomalyDetectionModels: Array[MetricAnomalyModel] = Array[MetricAnomalyModel]()

  def main(args: Array[String]): Unit = {

    @transient
    lazy val log: Logger = org.apache.log4j.LogManager.getLogger("MetricAnomalyDetectorLogger")

    if (args.length < 5) {
      System.err.println("Usage: MetricAnomalyDetector <method1,method2> <appid1,appid2> <collector_host> <port> <protocol>")
      System.exit(1)
    }

    for (method <- args(0).split(",")) {
      if (method == "ema") anomalyDetectionModels :+ new EmaModel()
    }

    val appIds = util.Arrays.asList(args(1).split(","))

    val collectorHost = args(2)
    val collectorPort = args(3)
    val collectorProtocol = args(4)

    val anomalyMetricPublisher: AnomalyMetricPublisher = new AnomalyMetricPublisher(collectorHost, collectorProtocol, collectorPort)

    val sparkConf = new SparkConf().setAppName("AmbariMetricsAnomalyDetector")

    val streamingContext = new StreamingContext(sparkConf, Duration(10000))

    val emaModel = new EmaModelLoader().load(streamingContext.sparkContext, "/tmp/model/ema")

    val kafkaStream = KafkaUtils.createStream(streamingContext, zkQuorum, groupId, Map(topicName -> numThreads), StorageLevel.MEMORY_AND_DISK_SER_2)
    kafkaStream.print()

    var timelineMetricsStream = kafkaStream.map( message => {
      val mapper = new ObjectMapper
      val metrics = mapper.readValue(message._2, classOf[TimelineMetrics])
      metrics
    })
    timelineMetricsStream.print()

    var appMetricStream = timelineMetricsStream.map( timelineMetrics => {
      (timelineMetrics.getMetrics.get(0).getAppId, timelineMetrics)
    })
    appMetricStream.print()

    var filteredAppMetricStream = appMetricStream.filter( appMetricTuple => {
      appIds.contains(appMetricTuple._1)
    } )
    filteredAppMetricStream.print()

    filteredAppMetricStream.foreachRDD( rdd => {
      rdd.foreach( appMetricTuple => {
        val timelineMetrics = appMetricTuple._2
        logger.info("Received Metric (1): " + timelineMetrics.getMetrics.get(0).getMetricName)
        log.info("Received Metric (2): " + timelineMetrics.getMetrics.get(0).getMetricName)
        for (timelineMetric <- timelineMetrics.getMetrics) {
          var anomalies = emaModel.test(timelineMetric)
          anomalyMetricPublisher.publish(anomalies)
          for (anomaly <- anomalies) {
            var an = anomaly : MetricAnomaly
            logger.info(an.getAnomalyAsString)
          }
        }
      })
    })

    streamingContext.start()
    streamingContext.awaitTermination()
  }
  }
